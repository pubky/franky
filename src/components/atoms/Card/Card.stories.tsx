import type { Meta, StoryObj } from '@storybook/nextjs-vite';

import {
  Card,
  CardHeader,
  CardTitle,
  CardDescription,
  CardContent,
  CardFooter,
  CardAction,
} from '@/components/atoms/Card';
import { Button } from '@/components/atoms/Button';

const meta = {
  title: 'Atoms/Card',
  component: Card,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Card>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: () => (
    <Card className="w-80">
      <CardHeader>
        <CardTitle>Card Title</CardTitle>
        <CardDescription>Card description goes here</CardDescription>
      </CardHeader>
      <CardContent>
        <p>This is the card content area where you can put any content.</p>
      </CardContent>
      <CardFooter>
        <Button variant="outline">Cancel</Button>
        <Button>Save</Button>
      </CardFooter>
    </Card>
  ),
};

export const WithAction: Story = {
  render: () => (
    <Card className="w-80">
      <CardHeader>
        <CardTitle>Card with Action</CardTitle>
        <CardDescription>This card has an action button in the header</CardDescription>
        <CardAction>
          <Button size="sm" variant="ghost">
            ⚙️
          </Button>
        </CardAction>
      </CardHeader>
      <CardContent>
        <p>Content with header action</p>
      </CardContent>
    </Card>
  ),
};

export const Simple: Story = {
  render: () => (
    <Card className="w-80">
      <CardContent>
        <p>Simple card with just content</p>
      </CardContent>
    </Card>
  ),
};

export const WithBorder: Story = {
  render: () => (
    <Card className="w-80 border">
      <CardHeader className="border-b">
        <CardTitle>Bordered Card</CardTitle>
        <CardDescription>This card has borders</CardDescription>
      </CardHeader>
      <CardContent>
        <p>Content with borders</p>
      </CardContent>
      <CardFooter className="border-t">
        <Button variant="outline">Action</Button>
      </CardFooter>
    </Card>
  ),
};
