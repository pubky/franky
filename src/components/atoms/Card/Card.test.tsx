import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Card, CardHeader, CardTitle, CardDescription, CardContent, CardFooter, CardAction } from './Card';

describe('Card', () => {
  it('renders with default props', () => {
    render(<Card data-testid="card">Card content</Card>);
    const card = screen.getByTestId('card');
    expect(card).toBeInTheDocument();
    expect(card).toHaveClass(
      'bg-card',
      'text-card-foreground',
      'flex',
      'flex-col',
      'gap-6',
      'rounded-xl',
      'py-6',
      'shadow-sm',
    );
  });

  it('applies custom className', () => {
    render(
      <Card data-testid="card" className="custom-class">
        Card content
      </Card>,
    );
    const card = screen.getByTestId('card');
    expect(card).toHaveClass('custom-class');
  });

  it('renders card header correctly', () => {
    render(<CardHeader data-testid="card-header">Header content</CardHeader>);
    const header = screen.getByTestId('card-header');
    expect(header).toBeInTheDocument();
    expect(header).toHaveAttribute('data-slot', 'card-header');
  });

  it('renders card title correctly', () => {
    render(<CardTitle data-testid="card-title">Title</CardTitle>);
    const title = screen.getByTestId('card-title');
    expect(title).toBeInTheDocument();
    expect(title).toHaveAttribute('data-slot', 'card-title');
    expect(title).toHaveClass('leading-none', 'font-semibold');
  });

  it('renders card description correctly', () => {
    render(<CardDescription data-testid="card-description">Description</CardDescription>);
    const description = screen.getByTestId('card-description');
    expect(description).toBeInTheDocument();
    expect(description).toHaveAttribute('data-slot', 'card-description');
    expect(description).toHaveClass('text-muted-foreground', 'text-sm');
  });

  it('renders card content correctly', () => {
    render(<CardContent data-testid="card-content">Content</CardContent>);
    const content = screen.getByTestId('card-content');
    expect(content).toBeInTheDocument();
    expect(content).toHaveAttribute('data-slot', 'card-content');
    expect(content).toHaveClass('px-6');
  });

  it('renders card footer correctly', () => {
    render(<CardFooter data-testid="card-footer">Footer</CardFooter>);
    const footer = screen.getByTestId('card-footer');
    expect(footer).toBeInTheDocument();
    expect(footer).toHaveAttribute('data-slot', 'card-footer');
    expect(footer).toHaveClass('flex', 'items-center', 'px-6');
  });

  it('renders card action correctly', () => {
    render(<CardAction data-testid="card-action">Action</CardAction>);
    const action = screen.getByTestId('card-action');
    expect(action).toBeInTheDocument();
    expect(action).toHaveAttribute('data-slot', 'card-action');
  });

  it('renders complete card structure', () => {
    render(
      <Card data-testid="complete-card">
        <CardHeader>
          <CardTitle>Card Title</CardTitle>
          <CardDescription>Card Description</CardDescription>
          <CardAction>Action</CardAction>
        </CardHeader>
        <CardContent>Card Content</CardContent>
        <CardFooter>Card Footer</CardFooter>
      </Card>,
    );

    expect(screen.getByTestId('complete-card')).toBeInTheDocument();
    expect(screen.getByText('Card Title')).toBeInTheDocument();
    expect(screen.getByText('Card Description')).toBeInTheDocument();
    expect(screen.getByText('Action')).toBeInTheDocument();
    expect(screen.getByText('Card Content')).toBeInTheDocument();
    expect(screen.getByText('Card Footer')).toBeInTheDocument();
  });
});

describe('Card - Snapshots', () => {
  it('matches snapshot with default props', () => {
    const { container } = render(<Card>Card content</Card>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshots for different configurations', () => {
    const { container: defaultContainer } = render(<Card>Default Card</Card>);
    expect(defaultContainer.firstChild).toMatchSnapshot();

    const { container: customContainer } = render(<Card className="custom-card">Custom Card</Card>);
    expect(customContainer.firstChild).toMatchSnapshot();
  });

  it('matches snapshots for card subcomponents', () => {
    const { container: headerContainer } = render(<CardHeader>Header Content</CardHeader>);
    expect(headerContainer.firstChild).toMatchSnapshot();

    const { container: titleContainer } = render(<CardTitle>Card Title</CardTitle>);
    expect(titleContainer.firstChild).toMatchSnapshot();

    const { container: descriptionContainer } = render(<CardDescription>Card Description</CardDescription>);
    expect(descriptionContainer.firstChild).toMatchSnapshot();

    const { container: contentContainer } = render(<CardContent>Card Content</CardContent>);
    expect(contentContainer.firstChild).toMatchSnapshot();

    const { container: footerContainer } = render(<CardFooter>Card Footer</CardFooter>);
    expect(footerContainer.firstChild).toMatchSnapshot();

    const { container: actionContainer } = render(<CardAction>Action</CardAction>);
    expect(actionContainer.firstChild).toMatchSnapshot();
  });

  it('matches snapshots for complete card structure', () => {
    const { container: completeContainer } = render(
      <Card>
        <CardHeader>
          <CardTitle>Card Title</CardTitle>
          <CardDescription>Card Description</CardDescription>
          <CardAction>Action</CardAction>
        </CardHeader>
        <CardContent>Card Content</CardContent>
        <CardFooter>Card Footer</CardFooter>
      </Card>,
    );
    expect(completeContainer.firstChild).toMatchSnapshot();

    const { container: minimalContainer } = render(
      <Card>
        <CardContent>Minimal Content</CardContent>
      </Card>,
    );
    expect(minimalContainer.firstChild).toMatchSnapshot();
  });
});
