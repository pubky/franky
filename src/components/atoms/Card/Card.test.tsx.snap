// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card - Snapshots > matches snapshot for CardAction 1`] = `
<div
  class="col-start-2 row-span-2 row-start-1 self-start justify-self-end"
  data-slot="card-action"
>
  Action
</div>
`;

exports[`Card - Snapshots > matches snapshot for CardContent 1`] = `
<div
  class="px-6"
  data-slot="card-content"
>
  Card Content
</div>
`;

exports[`Card - Snapshots > matches snapshot for CardDescription 1`] = `
<div
  class="text-muted-foreground text-sm"
  data-slot="card-description"
>
  Card Description
</div>
`;

exports[`Card - Snapshots > matches snapshot for CardFooter 1`] = `
<div
  class="flex items-center px-6 [.border-t]:pt-6"
  data-slot="card-footer"
>
  Card Footer
</div>
`;

exports[`Card - Snapshots > matches snapshot for CardHeader 1`] = `
<div
  class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
  data-slot="card-header"
>
  Header Content
</div>
`;

exports[`Card - Snapshots > matches snapshot for CardTitle 1`] = `
<div
  class="leading-none font-semibold"
  data-slot="card-title"
>
  Card Title
</div>
`;

exports[`Card - Snapshots > matches snapshot for complete card structure 1`] = `
<div
  class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl py-6 shadow-sm"
  data-slot="card"
  data-testid="card"
>
  <div
    class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
    data-slot="card-header"
  >
    <div
      class="leading-none font-semibold"
      data-slot="card-title"
    >
      Card Title
    </div>
    <div
      class="text-muted-foreground text-sm"
      data-slot="card-description"
    >
      Card Description
    </div>
    <div
      class="col-start-2 row-span-2 row-start-1 self-start justify-self-end"
      data-slot="card-action"
    >
      Action
    </div>
  </div>
  <div
    class="px-6"
    data-slot="card-content"
  >
    Card Content
  </div>
  <div
    class="flex items-center px-6 [.border-t]:pt-6"
    data-slot="card-footer"
  >
    Card Footer
  </div>
</div>
`;

exports[`Card - Snapshots > matches snapshot for minimal card structure 1`] = `
<div
  class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl py-6 shadow-sm"
  data-slot="card"
  data-testid="card"
>
  <div
    class="px-6"
    data-slot="card-content"
  >
    Minimal Content
  </div>
</div>
`;

exports[`Card - Snapshots > matches snapshot with custom className 1`] = `
<div
  class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl py-6 shadow-sm custom-card"
  data-slot="card"
  data-testid="card"
>
  Custom Card
</div>
`;

exports[`Card - Snapshots > matches snapshot with default props 1`] = `
<div
  class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl py-6 shadow-sm"
  data-slot="card"
  data-testid="card"
>
  Card content
</div>
`;
